File: backend/lib/build_info.py
"""Which build is actually running — resolved from the environment only.

Why this module exists: the only build identity the API published was
``API_VERSION``, a hand-maintained string that has not been bumped since the
commit that introduced it, so ``"2.7.0"`` says nothing about what code is
serving a request. The single place the real commit appeared was a Sentry
release tag (``backend/lib/observability.py``), and Sentry is DSN-gated —
with no DSN configured the commit was recorded nowhere at all. There was no
way to ask a running instance what SHA it was, which means no way to tell
whether a deploy landed, whether a rollback took effect, or which code
produced a bad response.

Resolution order for the SHA:

1. ``RENDER_GIT_COMMIT`` — injected by Render for the commit it deployed.
   The host's own statement about what it checked out is the strongest
   evidence available from inside the process.
2. ``OFE_RELEASE_SHA`` — an explicit operator override for hosts that do not
   announce the commit themselves (Docker, Fly, a local prod-like run) and
   for the frontend/CI plumbing that shares the same variable name.
3. Nothing → ``None``.

What this module deliberately does NOT do:

* It never synthesizes a placeholder ("dev", "local", "unknown-sha",
  "0000000"). A fabricated SHA is strictly worse than no SHA: it looks like
  provenance while proving nothing, and it would silently satisfy exactly
  the audit this module answers.
* It never shells out to ``git rev-parse``. The deployed artifact has no
  ``.git`` directory (Render builds from a source tarball), so the call
  would fail in the only environment where the answer matters; and where it
  *would* succeed — a developer box, a CI checkout — it reports whatever
  happens to be on disk, not the code this process loaded. That is
  provenance theater, so the fallback is honest ``None`` instead.
* It publishes nothing but build identity. ``build_info`` is served on an
  unauthenticated endpoint, so it may never grow into an environment dump:
  every field below is either a commit SHA, a hard-coded version, a
  host-name label, or this process's start time.

Unknown is represented as ``None`` for the SHA (JSON ``null``) and the
literal string ``"unknown"`` for the environment. Callers must render those
as unknown — never as "fresh", "current", or "production".
"""

from __future__ import annotations

import os
import re
from datetime import UTC, datetime

# The declared API contract version. It describes the shape of the API, not
# the deployed artifact — ``release_sha`` is the only field that identifies
# code. Kept here so ``backend.main`` re-exports one source of truth.
BUILD_VERSION = "2.7.0"

UNKNOWN_ENVIRONMENT = "unknown"

# Environment variables consulted for the commit, in order of trust.
SHA_ENV_VARS: tuple[str, ...] = ("RENDER_GIT_COMMIT", "OFE_RELEASE_SHA")

_SHORT_SHA_LENGTH = 7

# A git commit is hex. Anything else — an empty string, the word "unknown",
# an unexpanded "$RENDER_GIT_COMMIT" template, a branch name — is not a
# build identity and must not be published as one.
_SHA_RE = re.compile(r"^[0-9a-fA-F]{7,40}$")

# Captured once, at import: the moment this process came up. Recomputing it
# per request would report "now" and hide how long the instance has been
# running (and therefore whether a deploy actually restarted it).
_STARTED_AT = datetime.now(UTC)


def _clean(name: str) -> str:
    return (os.environ.get(name) or "").strip()


def release_sha() -> str | None:
    """The full commit SHA of the running build, or ``None`` if unknown."""
    for name in SHA_ENV_VARS:
        candidate = _clean(name)
        if _SHA_RE.match(candidate):
            return candidate.lower()
    return None


def release_sha_short() -> str | None:
    """The first 7 characters of :func:`release_sha`, or ``None``."""
    full = release_sha()
    return full[:_SHORT_SHA_LENGTH] if full else None


def build_version() -> str:
    """The declared API contract version (see :data:`BUILD_VERSION`)."""
    return BUILD_VERSION


def environment() -> str:
    """Where this process is running: a host label, or ``"unknown"``.

    Render sets ``RENDER=true`` in every runtime it starts, so that is the
    ground truth about the host and is consulted first; ``OFE_ENVIRONMENT``
    is the label for hosts that do not identify themselves. Neither is
    interpreted as "production" — this returns what the environment stated,
    never an inference.
    """
    if _clean("RENDER").lower() in {"1", "true", "yes"}:
        return "render"
    return _clean("OFE_ENVIRONMENT") or UNKNOWN_ENVIRONMENT


def started_at() -> str:
    """ISO-8601 UTC timestamp of this process's start."""
    return _STARTED_AT.isoformat()


def health_build_fields() -> dict[str, object]:
    """The build-provenance fields added to ``/api/health``.

    Exactly three keys, all safe to serve unauthenticated. ``status`` and
    ``version`` stay owned by the handler because three existing consumers
    (frontend ``wakeBackend``, the Playwright ``webServer`` readiness gate,
    and ``tests/test_async_route_isolation``) depend on their current shape.
    """
    return {
        "release_sha": release_sha(),
        "environment": environment(),
        "started_at": started_at(),
    }


def build_info() -> dict[str, object]:
    """The complete build identity record, unknowns included."""
    return {
        "release_sha": release_sha(),
        "release_sha_short": release_sha_short(),
        "build_version": build_version(),
        "environment": environment(),
        "started_at": started_at(),
    }

File: frontend/src/lib/build-info.ts
/**
 * Which frontend build is being served.
 *
 * The client shipped no build identity at all: `VERCEL_GIT_COMMIT_SHA` was
 * never read, so a page in front of a user could not be traced back to a
 * commit — "is the fix deployed?" was unanswerable except by guessing from
 * behaviour. `next.config.js` now inlines the commit at build time as
 * `NEXT_PUBLIC_RELEASE_SHA` (Vercel's `VERCEL_GIT_COMMIT_SHA`, falling back
 * to an explicit `OFE_RELEASE_SHA`), and this module reads it.
 *
 * Honest-unknown rule, identical to `backend/lib/build_info.py`:
 * unknown is `null`, never a placeholder. A value that is missing, empty, or
 * not shaped like a commit (an unexpanded `$VERCEL_GIT_COMMIT_SHA`, the word
 * "unknown", a branch name) is reported as unknown rather than published as
 * provenance. A fabricated SHA would look like proof and be none.
 *
 * There is deliberately no `git rev-parse` fallback: the value must describe
 * the deployed artifact, and a dev-box checkout says nothing about it.
 */

const SHA_PATTERN = /^[0-9a-f]{7,40}$/i;

export const UNKNOWN_ENVIRONMENT = 'unknown';

const SHORT_SHA_LENGTH = 7;

/** A commit SHA, or `null` when the input is absent or not SHA-shaped. */
export function normalizeSha(raw: string | undefined | null): string | null {
  const value = (raw ?? '').trim();
  return SHA_PATTERN.test(value) ? value.toLowerCase() : null;
}

/** The provided environment label, or `"unknown"`. Never inferred. */
export function normalizeEnvironment(raw: string | undefined | null): string {
  return (raw ?? '').trim() || UNKNOWN_ENVIRONMENT;
}

export interface BuildInfo {
  releaseSha: string | null;
  releaseShaShort: string | null;
  environment: string;
}

/** Pure resolver — exported so the contract is testable without a rebuild. */
export function resolveBuildInfo(
  env: { NEXT_PUBLIC_RELEASE_SHA?: string; NEXT_PUBLIC_RELEASE_ENV?: string },
): BuildInfo {
  const sha = normalizeSha(env.NEXT_PUBLIC_RELEASE_SHA);
  return {
    releaseSha: sha,
    releaseShaShort: sha ? sha.slice(0, SHORT_SHA_LENGTH) : null,
    environment: normalizeEnvironment(env.NEXT_PUBLIC_RELEASE_ENV),
  };
}

const buildInfo = resolveBuildInfo({
  NEXT_PUBLIC_RELEASE_SHA: process.env.NEXT_PUBLIC_RELEASE_SHA,
  NEXT_PUBLIC_RELEASE_ENV: process.env.NEXT_PUBLIC_RELEASE_ENV,
});

export const releaseSha = buildInfo.releaseSha;
export const releaseShaShort = buildInfo.releaseShaShort;
export const environment = buildInfo.environment;

/**
 * What to put in the DOM. `"unknown"` (not an omitted attribute and not a
 * fake SHA) so that an operator reading production HTML can tell "the
 * plumbing works but no SHA was supplied" apart from "this build predates
 * release stamping".
 */
export const releaseShaAttribute = releaseSha ?? UNKNOWN_ENVIRONMENT;
